### PR TITLE
[pal] Add api to allow setting the outer bounds for a label feature 

### PR DESCRIPTION
**Pull request by nyalldawson**
The outer bounds can represent the extreme margins of a label, including
non-text parts like borders or padded backgrounds. During the label
solving the outer bounds will be used when detecting whether
labels are in conflict.

### CODE REVIEW

File: src/core/labeling/qgsvectorlayerlabelprovider.cpp [hunk 1]
```diff
@@ -569,6 +602,8 @@ void QgsVectorLayerLabelProvider::drawLabelPrivate( pal::LabelPosition *label, Q
         painter->drawLine( QPointF( rect.left(), rect.top()  + blockBaseLine ), QPointF( rect.right(), rect.top() + blockBaseLine ) );
       }
     }
+
+
```
Comment (nirvn) [inline src/core/labeling/qgsvectorlayerlabelprovider.cpp:606]:
```suggestion
```

File: src/core/pal/labelposition.h [hunk 2]
```diff
@@ -97,55 +97,62 @@ namespace pal
...
+       * Returns TRUE if the outer bounding box of this pointset intersects the outer bounding box
+       * of another label position.
+       */
+      bool outerBoundingBoxIntersects( const LabelPosition *other ) const;
 
-      //! Returns TRUE if this label crosses the specified line
+      /**
+       * Gets distance from this label to a point. If point lies inside, returns negative number.
+       */
+      double getDistanceToPoint( double xp, double yp, bool useOuterBounds ) const;
```
Comment (nirvn) [inline src/core/pal/labelposition.h:142]: Could you document what useOuterBounds = true vs useOuterBounds = false does?

File: src/core/pal/labelposition.cpp [hunk 3]
```diff
@@ -457,44 +539,116 @@ const GEOSPreparedGeometry *LabelPosition::preparedMultiPartGeom() const
...
 
-double LabelPosition::getDistanceToPoint( double xp, double yp ) const
+const GEOSPreparedGeometry *LabelPosition::preparedOuterBoundsGeom() const
+{
+  return mPreparedOuterBoundsGeos;
+}
+
+double LabelPosition::getDistanceToPoint( double xp, double yp, bool useOuterBounds ) const
 {
+  // this method considers the label's outer bounds
```
Comment (nirvn) [inline src/core/pal/labelposition.cpp:549]: This method _may_ consider the label's outer bounds?

**Review by nirvn — COMMENTED** [review]
Excited for what comes next. Some comments left.

**Comment by nyalldawson** [comment]
Thanks @nirvn!

**Review by nirvn — APPROVED** [review]

**Comment by nirvn** [comment]
@nyalldawson , approved.

**Pull request merged**